src/Lib.P2P/SwarmService.cs

```diff
@@ -0,0 +1,1132 @@
+        /// </summary>
+        /// <param name="peer">
+        ///   The peer to remove..
+        /// </param>
+        /// <remarks>
+        ///   Remove all knowledge of the peer. The <see cref="PeerRemoved"/> event
+        ///   is raised.
+        /// </remarks>
+        public void DeregisterPeer(Peer peer)
+        {
+            if (peer.Id == null)
+            {
+                throw new ArgumentNullException(nameof(peer));
+            }
+
+            if (_otherPeers.TryRemove(peer.Id.ToBase58(), out var found))
+            {
+                peer = found;
+            }
+            
+            PeerRemoved?.Invoke(this, peer);
+        }
+
+        /// <summary>
+        ///   Determines if a connection is being made to the peer.
+        /// </summary>
+        /// <param name="peer">
+        ///   A <see cref="Peer"/>.
+        /// </param>
+        /// <returns>
+        ///   <b>true</b> is the <paramref name="peer"/> has a pending connection.
+        /// </returns>
+        public bool HasPendingConnection(Peer peer) { return _pendingConnections.TryGetValue(peer, out _); }
+
+        /// <summary>
+        ///   The addresses that cannot be used.
+        /// </summary>
+        public MultiAddressBlackList BlackList { get; set; } = new MultiAddressBlackList();
+
+        /// <summary>
+        ///   The addresses that can be used.
+        /// </summary>
+        public MultiAddressWhiteList WhiteList { get; set; } = new MultiAddressWhiteList();
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="dhtService"></param>
+        public SwarmService(IPeerRouting dhtService = null) { Router = dhtService; }
+
+        /// <inheritdoc />
+        public Task StartAsync()
+        {
+            if (LocalPeer == null)
+            {
+                throw new NotSupportedException("The LocalPeer is not defined.");
+            }
+
+            // Many of the unit tests do not setup the LocalPeerKey.  If
+            // its missing, then just use plaintext connection.
+            // TODO: make the tests setup the security protocols.
+            if (LocalPeerKey == null)
+            {
+                lock (_protocols)
+                {
+                    var security = _protocols.OfType<IEncryptionProtocol>().ToArray();
+                    foreach (var p in security)
+                    {
+                        _protocols.Remove(p);
+                    }
+                    
+                    _protocols.Add(_plaintext1);
+                }
+
+                _log.Warn("Peer key is missing, using unencrypted connections.");
+            }
+
+            Manager.PeerDisconnected += OnPeerDisconnected;
+            IsRunning = true;
+            _swarmCancellation = new CancellationTokenSource();
+            _log.Debug("Started");
+
+            return Task.CompletedTask;
+        }
+
+        private void OnPeerDisconnected(object sender, MultiHash peerId)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this unused method parameter 'sender'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=5051852)